python-spec/src/somabase/_wrap.py

```diff
@@ -0,0 +1,220 @@
+
+    def __setitem__(self, key: str, value: _ST) -> None:
+        self._backing[key] = value
+
+    def __delitem__(self, key: str) -> None:
+        del self._backing[key]
+
+    def pop(
+        self,
+        key: str,
+        value: Union[_T, _ST] = _SENTINEL,  # type: ignore[assignment]
+    ) -> Union[_T, _ST]:
+        if value is _SENTINEL:
+            return self._backing.pop(key)
+        return self._backing.pop(key, value)
+
+    def popitem(self) -> Tuple[str, _ST]:
+        return self._backing.popitem()
+
+    def clear(self) -> None:
+        self._backing.clear()
+
+    def update(self, __other=(), **kwds) -> None:
+        return self._backing.update(__other, **kwds)
+
+    def setdefault(
+        self,
+        key: str,
+        default: _ST = None,  # type: ignore[assignment]
+    ) -> _ST:
+        return self._backing.setdefault(key, default)  # type: ignore[arg-type]
+
+
+@attrs.define()
+class item(Generic[_T]):
+    """Descriptor to transform property access into indexing.
+
+    This descriptor works with :class:`CollectionProxy` to allow for simple
+    specification of properties that reflect the members of the collection::
+
+        class WrapImpl(CollectionProxy):
+
+            first = item(str)
+            second = item(int, "2nd")
+
+        inst = WrapImpl(some_collection)
+
+        # This is equivalent to getting some_collection["first"]
+        inst.first
+
+        # This is equivalent to setting some_collection["2nd"]
+        inst.second = 500
+    """
+
+    typ: Optional[Type[_T]] = None
+    """The type we expect to return from this field."""
+
+    item_name: Optional[str] = None
+    """The name of the item we are getting (``x._backing["whatever"]``).
+
+    This uses the name of the field by default but can be manually overridden.
+    """
+
+    field_name: str = attrs.field(default="<unknown>", init=False)
+    """The name of this field (``x.whatever``). Set automatically."""
+
+    def __set_name__(self, owner: Type[CollectionProxy], name: str) -> None:
+        del owner  # unused
+        self.field_name = name
+        if self.item_name is None:
+            self.item_name = name
+
+    @overload
+    def __get__(self, inst: None, owner: Type[CollectionProxy]) -> "item[_T]":
+        ...
+
+    @overload
+    def __get__(self, inst: CollectionProxy, owner: Type[CollectionProxy]) -> _T:
+        ...
+
+    def __get__(
+        self, inst: Optional[CollectionProxy], owner: Type[CollectionProxy]
+    ) -> Union["item", _T]:
+        del owner  # unused
+        if not inst:
+            return self
+        assert self.item_name is not None
+        try:
+            # TODO: Type-check params/returns?
+            return inst[self.item_name]
+        except KeyError as ke:
+            raise AttributeError(
+                f"{_typename(inst)!r} object has no attribute {self.field_name!r}"
+            ) from ke
+
+    def __set__(self, inst: CollectionProxy, value: _T) -> None:
```

Comment:
Non-blocking question: down the road, if we need to implement additional constraint checks upon __set__ (e.g., perhaps check the shape of the value), how would this be extended or composed by the parent class (e.g., Measurement)?

Reply:
I think the way to do this would be to add a function similar to the way `@property` (itself a descriptor) works. This would look something like this:

```python

class item(...):
  # ...

  def validator(self, validator_func: Callable[...]):
    self._validator = validator
    return validator

  def __set__(self, inst, value):
    try:
      self._validator(value)
    except ...:
      raise ...
    # existing __set__ implementation
```

which would then be used like this:

```python
class SomeWrapper(_wrap.CollectionProxy):

  it = _wrap.item()

  @it.validator
  def _validate_it(self, new_it):
    if new_it.name in {"john", "paul", "george", "ringo"}:
      raise ValueError("no beatles allowed")
```

This might also be implemented as some kind of transform, where the value gets passed through a `it.transform_set` function or the like:

```python
  it = _wrap.item()

  @it.transform_set
  def _set_it(self, new_it):
    if new_it.name in ...:
      ...  # validation occurs as part of the transform step
    return do_something_to(new_it)
```

If an implementation needs to do its own custom validation or transformation, the value works just like any other property:

```python
class SpecialWrapper(some_mod.SomeWrapper):

  @property
  def it(self):
    return super().it

  @it.setter
  def _set_it(self, new_it):
    new_new_it = special_xform(new_it)
    super().it = new_new_it
```

Reply:
Great - I was thinking along similar lines.  Many thanks.